### PR TITLE
Handle `ReturnData` for other types

### PR DESCRIPTION
**Pull request by digorithm**
Closes https://github.com/FuelLabs/fuels-rs/issues/139. 

More context in https://github.com/FuelLabs/fuels-rs/issues/139. Summary: depending on the size of the returned data, the returned data will live either under the `Return` receipt or `ReturnData` receipt. This handles it for the types that weren't being properly handled before.

### CODE REVIEW

File: fuels-abigen-macro/tests/harness.rs
Comment (digorithm) [inline fuels-abigen-macro/tests/harness.rs:1105]: @nfurfaro I believe this is the relevant case for you: `get_contract_id()` returns the Sway type `ContractId`, which wraps the `b256` in a `Struct`, which wasn't being handled before. Now it seems to be working fine!

File: fuels-core/src/lib.rs
Comment (digorithm) [inline fuels-core/src/lib.rs:69]: @vnepveu I believe this will be related to your current work on supporting multiple return values, make sure to rebase it on top of this once it's merged!

**Review by nfurfaro — COMMENTED** [review]
LGTM, but I'll defer to @vnepveu

File: fuels-abigen-macro/tests/test_projects/large-return-data/src/main.sw
Comment (adlerjohn) [inline fuels-abigen-macro/tests/test_projects/large-return-data/src/main.sw:24]: https://github.com/FuelLabs/sway/issues/726 strikes again. cc @eureka-cpu

File: fuels-core/src/lib.rs
Comment (adlerjohn) [inline fuels-core/src/lib.rs:48]:
```suggestion
    // This is important because, depending on whether it's
```

File: fuels-core/src/lib.rs
Comment (adlerjohn) [inline fuels-core/src/lib.rs:65]: Is this expected or a bug? cc @otrho

Comment (otrho) [reply fuels-core/src/lib.rs:65]: I'm guessing at this stage it's always true since even an enum with a single variant will still have a tag and a value, even if it's just `unit`.  So it can't fit in a register.  After we add some gnarly optimisations this may no longer be true, but only if we want to go there.

Comment (adlerjohn) [reply fuels-core/src/lib.rs:65]: Oh I was pointing at the ``Arrays seem to always be inside `ReturnData`.``, not the enums 😂

Comment (digorithm) [reply fuels-core/src/lib.rs:65]: Unfortunately I had to do a lot of guesswork to get this working and figure out where each returned type would be located at. 

As for arrays, in the tests I did, they were always in ReturnData (there are few new array-as-output tests added in this PR). But then again, I didn't fuzz or anything, just tested a few cases and focused on solving some existing blockers. 

But, intuitively (which could be wrong of course), if any primitive type takes a whole word. Even if it doesn't fill it all we pad it. So an array should be larger than a word? Maybe unless is an array of size zero or one? 😅

File: fuels-core/src/lib.rs
Comment (vnepveu) [reply fuels-core/src/lib.rs:69]: Yes will do!

**Review by vnepveu — APPROVED** [review]
LGTM

File: fuels-core/src/lib.rs
Comment (adlerjohn) [reply fuels-core/src/lib.rs:65]: An array of size 0 or 1 should only take up a single word. Could you add a test for that @digorithm?

File: fuels-abigen-macro/tests/test_projects/large-return-data/src/main.sw
Comment (eureka-cpu) [reply fuels-abigen-macro/tests/test_projects/large-return-data/src/main.sw:24]: Likely what's happening is the formatter is adding a newline character when it finds a `;` and this edge case wasn't taken into account. Currently working on the `format stucts and enums` draft that I had yet to finish, so will take a look at this too.

**Review by adlerjohn — APPROVED** [review]
Can handle arrays of size 0 or 1 in a future PR. Make sure there's an issue for it.

**Comment by digorithm** [comment]
Tracking it here: https://github.com/FuelLabs/fuels-rs/issues/162.

**Pull request merged**
